### PR TITLE
Update Eslint JSDoc package, introducing JSDoc line alignment check

### LINKED ISSUE #24984: Eslint rule for JSDoc alignment
**Is your feature request related to a problem? Please describe.**

Currently, there is no Eslint plugin for check JSDoc alignment (at least I didn't find one in my researches 😁), as recommended in the [WordPress standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#aligning-comments).

**Describe the solution you'd like**

We created a new [plugin to check JSDoc alignments (eslint-plugin-jsdoc-alignment)](https://github.com/Automattic/eslint-plugin-jsdoc-alignment). As it can be useful by WordPress community in general, @nerrad gave a good idea to add it to the `@wordpress/eslint-plugin`. So, before proceeding, it'd be good to discuss the approach to be used for that.

Things to consider:

- It can be very helpful to the WordPress community, as it checks a WordPress standard not covered currently by `@wordpress/eslint-plugin`.
- It can also be used by other projects (without WordPress) that want to use JSDoc with the lines aligned.
- This plugin was designed to be used along with [eslint-plugin-jsdoc](https://www.npmjs.com/package/eslint-plugin-jsdoc) because it checks only the alignment between the lines.

**Describe alternatives you've considered**

1) We could add it to the [eslint-plugin/configs](https://github.com/WordPress/gutenberg/tree/master/packages/eslint-plugin/configs) installing the [current package](https://www.npmjs.com/package/eslint-plugin-jsdoc-alignment). In some moment we can also open a new PR for [eslint-plugin-jsdoc](https://www.npmjs.com/package/eslint-plugin-jsdoc) and when they accept, we could uninstall the package and just update the configuration to use the new rule from `eslint-plugin-jsdoc`.
2) We could migrate the rule to [eslint-plugin/rules](https://github.com/WordPress/gutenberg/tree/master/packages/eslint-plugin/rules) and add to the `eslint-plugin/configs`. I particularly don't like that, because it makes the plugin restrict only to `eslint-plugin`.
3) We could migrate the `eslint-plugin-jsdoc-alignment` to be a Gutenberg package (we could also make it for the other current Gutenberg generic rules), and install it in the `eslint-plugin/configs`. So devs that want use only some specific rule, can install it separately.

IMHO, the option 1 looks the more "organized", looking to the future with the contribution being accepted by `eslint-plugin-jsdoc`. It doesn't exclude the option to transform the other Gutenberg generic rules to Gutenberg packages.

### PR DESCRIPTION
**Pull request by renatho**
Fixes #24984

## Description

This PR updates the `eslint-plugin-jsdoc`, which now contains a rule to [validate JSDoc alignment](https://github.com/gajus/eslint-plugin-jsdoc/pull/636), as recommended in [WordPress standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#aligning-comments).

## How has this been tested?

Eslint checking in the Gutenberg code.

## Screenshots 

Screen Shot 2020-09-14 at 12 08 51

## Types of changes

It's about Eslint rule updates. As it updates the rules, some new Eslint warnings/errors can appear considering the new rule added and updates in the plugin.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by renatho** [comment]
I didn't fix in this PR the new reported Gutenbeg Eslint issues. How should we proceed in cases like that when we update Eslint dependencies?

Maybe the cleaner way could be creating another PR with the fixes and merge before this PR. If it makes sense, I can create this PR.

**Comment by ntwb** [comment]
> I didn't fix the new reported Gutenbeg Eslint issues in this PR. How should we proceed in cases like that when we update Eslint dependencies?

> Maybe the cleaner way could be creating another PR with the fixes and merge it before this PR. If it makes sense, I can create this PR.

This I think makes sense? Is there many instances to fix?

**Comment by renatho** [comment]
> This I think makes sense? Is there many instances to fix?

Yep! There were a lot of issues to fix. But the most part was automatically updated.
I created the separate PR #25352 for this purpose.

I also updated the new rule to `warning` instead of `error` here: https://github.com/WordPress/gutenberg/pull/25300/commits/4191f713db4bac519adc7f5b5319f40b671313fb

I think it makes more sense as it's related to documentation format only.

**Comment by renatho** [comment]
> I also updated the new rule to warning instead of error here: 4191f71
> 
> I think it makes more sense as it's related to documentation format only.

It'd also allow us to merge before the other PR with the fixes.

**Comment by renatho** [comment]
I'd like to highlight this issue that @noahtallen found: https://github.com/WordPress/gutenberg/pull/25352#pullrequestreview-489158789

Just to check if someone thinks we should hold this change until we fix that in the plugin. In my opinion, I think it wouldn't be a big deal because it's easy to fix stuff.

**Comment by ntwb** [comment]
Looking good @renatho, I've left some comments o the follow up PR

**Comment by ntwb** [comment]
I mentioned this on the PR, but I think this should be explored also:

It may in fact we worthwhile trying to use Prettier prettier-plugin-jsdoc here to update the docs s mentioned here in https://github.com/WordPress/gutenberg/issues/24984#issuecomment-690572969 as this may be signifantly quicker for both this PR and future code written will be formatted automatically rather than a devloper pushing up code that then fails linting

**Comment by renatho** [comment]
> I mentioned this on the PR, but I think this should be explored also:
> 
> It may in fact we worthwhile trying to use Prettier prettier-plugin-jsdoc here to update the docs s mentioned here in [#24984 (comment)](https://github.com/WordPress/gutenberg/issues/24984#issuecomment-690572969) as this may be signifantly quicker for both this PR and future code written will be formatted automatically rather than a devloper pushing up code that then fails linting

Hey @ntwb!

I totally agree! It's much easier while developing. I checked that plugin, but running against the Sensei LMS code, it threw some errors for me. I'm planning to take a look at this too and contribute with some PR with fixes if it makes sense.

**Comment by renatho** [comment]
> I'd like to highlight this issue that @noahtallen found: #25352 (review)
> 
> Just to check if someone thinks we should hold this change until we fix that in the plugin. In my opinion, I think it wouldn't be a big deal because it's easy to fix stuff.

I added a new PR with fixes to the Eslint plugin here https://github.com/gajus/eslint-plugin-jsdoc/pull/638

The maintainer suggested a different approach which I think makes sense. It'll just take a little more time, so I think I'll keep this PR (and the https://github.com/WordPress/gutenberg/pull/25352) in stand by until going with this fix. And in these changes, I can explore more the issues we found while formatting in the https://github.com/WordPress/gutenberg/pull/25352.

**Comment by gziolo** [comment]
@renatho, nice work on this PR. Looking forward to seeing future revisions :)

**Comment by gziolo** [comment]
I believe it's still blocked by the pending patch in the ESLint plugin, right?

Is this rule fixable, in effect does it work with `eslint --fix`?

**Comment by renatho** [comment]
> I believe it's still blocked by the pending patch in the ESLint plugin, right?

Actually, now it's waiting for this change: https://github.com/syavorsky/comment-parser/issues/93.
It's a dependency of ESLint plugin, which is being refactored, and it'll include a `stringify` feature with a `format: "align"` property that can auto-align the whole JSDoc comment. After this is ready, the ESLint plugin will be updated, and then we can add it to Gutenberg. After that, we can also easily create a prettier rule using the same dependency.

> Is this rule fixable, in effect does it work with `eslint --fix`?

Yes! The current implementation is prepared to autofix, and after the changes it'll continue with this feature, but more complete than the current implementation.

The current implementation only checks and autofix the tags `'param', 'arg', 'argument', 'property', 'prop'`. And it wasn't checking if the multi-line descriptions were aligned. The new implementation will check and fix the whole thing. =)

**Comment by gziolo** [comment]
@renatho, super cool stuff. Thank you for your very detailed response 💯

**Comment by skorasaurus** [comment]
@gziolo @renatho
FYI: the ticket https://github.com/syavorsky/comment-parser/issues/93 was closed yesterday and 1.0 of comment-parser was released. Details inside that ticket.

**Comment by gziolo** [comment]
@skorasaurus, great news! I guess, we still need to wait until ESLint plugin gets updated as it's using 0.x version:

https://github.com/gajus/eslint-plugin-jsdoc/blob/2577f66a13eca9f8bac5895ccd0a634048dc5857/package.json#L8

**Comment by renatho** [comment]
@skorasaurus @gziolo Yep! Now the `eslint-plugin-jsdoc` needs to be updated to use the new Comment Parser, and we need to refactor the `jsdoc/check-line-alignment` rule to use it. =)

Very excited about that! As soon as I can I want to make refactor the rule.

**Comment by renatho** [comment]
Good news! `comment-parser` is released with their updates and `eslint-plugin-jsdoc` already updated the dependency! So now we just need to create a PR to `eslint-plugin-jsdoc` updating the rule to use it! I wanna implement it as soon as I can!

File: packages/rich-text/src/create.js
Comment (renatho) [inline packages/rich-text/src/create.js:116]: This is the only case where I disabled the rule because adding an `@example` tag would add a _Usage_ mark in the readme.

File: packages/blocks/src/api/validation/index.js
Comment (renatho) [inline packages/blocks/src/api/validation/index.js:47]: This, like the following cases, is not exactly an example of the content being used. But I think can interpret it as an example of the code that was used to extract it. Does it make sense?

File: packages/rich-text/src/create.js
Comment (renatho) [inline packages/rich-text/src/create.js]: This is the only case where I disabled the rule because adding an @example tag would add a Usage mark in the readme.

File: packages/blocks/src/api/validation/index.js
Comment (renatho) [inline packages/blocks/src/api/validation/index.js]: This, like the following cases, is not exactly an example of the content being used. But I think can interpret it as an example of the code that was used to extract it. Does it make sense?

File: packages/eslint-plugin/configs/jsdoc.js
Comment (renatho) [inline packages/eslint-plugin/configs/jsdoc.js:107]: Since this rule started being validated after the `eslint-plugin-jsdoc` update, I turned it off because there's many cases it was not adding the `@yield`.

**Review by renatho — COMMENTED** [review]
After a long time 🥵, it's ready for review. 🥳

Besides the [simpler jsdoc](https://github.com/WordPress/gutenberg/pull/25300/commits/acbbd3ca7824c0c64526766fa0a4a2a46e11a548) formatting, we had some [questionable formatting](https://github.com/WordPress/gutenberg/pull/25300/commits/fc159a50edab9d6c54ef003f581f24dae4e0701c). So I separated them into that 2 different commits in order to make the review easier. Basically, the "questionable" formatting doesn't apply for many cases, and I think the workaround is okay. But if we don't think it's okay, we can hold it for some more time and try to support a new configuration to preserve the left spacings of the text in JSDoc descriptions.

I added some comments to the code to illustrate the cases.

I was checking how to implement it, and it'll likely be a little hard, being necessary to implement changes in some different dependencies.

**Update:** Just one more observation. I added the rule aligning the default tags: `'param', 'arg', 'argument', 'property', 'prop', 'returns', 'return'`. In a few cases, I noticed we had an example between the `@param` and `@return`. I didn't change it, but it'd more beautiful if we add the `@example` before or after the aligned tags. If we don't like that, we can just change the order, or we could also not align the `@return`, although I think it's nice to keep it aligned too.

File: packages/rich-text/src/create.js
Comment (kevin940726) [reply packages/rich-text/src/create.js]: What if we modify the rule to skip formatting if we encounter a markdown code block? So that we don't have to use `@example` whenever we want to insert some formatted code.

File: bin/api-docs/update-api-docs.js
Comment (kevin940726) [inline bin/api-docs/update-api-docs.js]: Maybe I'm missing something, but I don't get why we have to add these spaces here?

**Review by kevin940726 — COMMENTED** [review]
Just found this gem! Super cool stuff!

File: bin/api-docs/update-api-docs.js
Comment (swissspidy) [reply bin/api-docs/update-api-docs.js]: Sounds like the rule aligns both `@param` _and_ `@return` lines, not just `@param` as the coding standards suggest. Although the docs are not super clear about it, tbh. But in WP core the `@return` lines aren't usually aligned.

I think the rule would need to be made configurable to allow for this.

Comment (gziolo) [reply bin/api-docs/update-api-docs.js]: Right, it would be great to exclude `@return` from this formatting 👍🏻

**Comment by gziolo** [comment]
@renatho, many thanks for keeping a pulse on this PR for so long. It looks very close, we mostly need to agree on the final shape of formatting 🎉

**Comment by swissspidy** [comment]
> **Update:** Just one more observation. I added the rule aligning the default tags: `'param', 'arg', 'argument', 'property', 'prop', 'returns', 'return'`. In a few cases, I noticed we had an example between the `@param` and `@return`. I didn't change it, but it'd more beautiful if we add the `@example` before or after the aligned tags. If we don't like that, we can just change the order, or we could also not align the `@return`, although I think it's nice to keep it aligned too.

@renatho Thanks for working on this! The `@return` lines are never aligned in WP core code and coding standards examples, so let's exclude them.

**Comment by renatho** [comment]
> @renatho, many thanks for keeping a pulse on this PR for so long. It looks very close, we mostly need to agree on the final shape of formatting 🎉

> @renatho Thanks for working on this!

My pleasure! =)

> The `@return` lines are never aligned in WP core code and coding standards examples, so let's exclude them.

Sure!

I'm just wondering about the https://github.com/WordPress/gutenberg/pull/25300#discussion_r629795857 ([related changes](https://github.com/WordPress/gutenberg/pull/25300/commits/fc159a50edab9d6c54ef003f581f24dae4e0701c)). Do you think it's really important and we should hold on until we have a solution for that? Or since it's not many cases we could go ahead with the current workaround?

**Comment by renatho** [comment]
Just an update that I'm creating a new PR to `eslint-plugin-jsdoc` adding an option to preserve the spaces in the main description. ;)

I thought we'd need to change other dependencies, but it wasn't necessary.

https://github.com/gajus/eslint-plugin-jsdoc/pull/731

File: packages/rich-text/src/create.js
Comment (kevin940726) [reply packages/rich-text/src/create.js]: > Do you think it's really important and we should hold on until we have a solution for that? Or since it's not many cases we could go ahead with the current workaround?

No, I don't think it's that important. Just tossing some ideas for potential improvements :)

**Comment by renatho** [comment]
Done! =)

Now it [allows spacings](https://github.com/WordPress/gutenberg/pull/25300/commits/c2ba0cd43d37b46657411b169493884c614d06e6) in the main JSDoc description.

And the rule just check alignment for the tags `'param', 'arg', 'argument', 'property', 'prop'`.

I manually [removed](https://github.com/WordPress/gutenberg/pull/25300/commits/3ce2d3f8cfbcde9b4d9ef255e0d756c2463f0404) the alignment from some cases that were aligning the `@return` (In the future it could contain some customization to remove the extra spacings for the tags not being aligned).

Just a notice that I got some unrelated eslint errors: Unable to resolve path to module '@wordpress/widgets'. So I needed to commit with --no-verify to skip these errors when I identified that it was the only errors. ;)

**Comment by renatho** [comment]
Hey folks! Just to let you know that I'll wait for review before fixing the conflicts. Just because it affects many files and the conflicts will appear very often. ;)

About the unit test failing, it seems unrelated. I also noticed it in another PR. I imagine when I rebase it again, probably it will be fixed.

File: packages/docgen/lib/get-type-annotation.js
Comment (swissspidy) [inline packages/docgen/lib/get-type-annotation.js]: Not the best readability 😅 

Perhaps the `typeAnnotation` type could be moved to a `@typedef`? That would help improve alignment.

File: packages/eslint-plugin/configs/jsdoc.js
Comment (swissspidy) [inline packages/eslint-plugin/configs/jsdoc.js:110]: This new addition seems to warrant a changelog entry

File: packages/docgen/lib/get-type-annotation.js
Comment (renatho) [reply packages/docgen/lib/get-type-annotation.js]: Sure! Makes sense! I updated it here: https://github.com/WordPress/gutenberg/pull/25300/commits/56b0a2c1f1504aeba0622e041c8890200c2d7884

Since we already had a `TypeAnnotation` typedef, I named it `ExtendedTypeAnnotation`. Do you think it makes sense here?

I think we have a few other similar cases in the code that would demand manual improvements, but I tried not to fix all of them, since they also need some context understanding.

File: packages/eslint-plugin/configs/jsdoc.js
Comment (renatho) [reply packages/eslint-plugin/configs/jsdoc.js:110]: Makes sense! Added here: https://github.com/WordPress/gutenberg/pull/25300/commits/2bf1f088acc6eeacc4de6f72e6b48b7379f508e0

**Comment by renatho** [comment]
I rebased the branch and fixed the issues again. Is there something I could do in order to go ahead with this PR to avoid more conflicts? 🙂

**Review by kevin940726 — APPROVED** [review]
This is looking great to me! 👍

As long as it doesn't break anything, I think we should merge this, and create follow-up PRs for other improvements if there's any.

**Comment by gziolo** [comment]
I don't see new warnings introduced with this PR so we are good:

Screen Shot 2021-06-09 at 09 22 42

As a follow-up, we should promote the reporting level from a warning to error for the new rule, but this is something for the future when people get a chance to update their codebases. One thing that we will have to look at is whether `@wordpress/create-block` generates JS files that follow the rule. At least the one that is tested with CI job (https://github.com/WordPress/gutenberg/pull/25300/checks?check_run_id=2778385774) looks good in that regard.

**Review by gziolo — APPROVED** [review]

**Comment by gziolo** [comment]
@renatho, congrats on your first merged pull request. It took some time but it required a ton of work so major props for all the time spent to make it happen 👏🏻 🎉

**Comment by renatho** [comment]
Hey folks! Thank you all for your reviews! And thank you for your patience in this very long PR! 😉

**Comment by hypest** [comment]
👋 friends! Over at the gutenberg-mobile repo, we seem to [run into check failures](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/12448/workflows/dae22fc0-48fe-4464-9aca-6151c5a51bf4/jobs/67087?invite=true#step-110-54) related to the JSDoc updates from this PR, and I wonder if anyone can give a quick hand or share insight. I'm sure we need to update [how we invoke the lint tests at this point](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/bin/ci-checks-js.sh#L35-L41) but any insight on what's the new proper way to do it will be appreciated! Thanks!

**Comment by gziolo** [comment]
@hypest, are you sure it's coming from this PR? There are no new errors added here, I would rather look at https://github.com/WordPress/gutenberg/pull/31792 first.

**Comment by renatho** [comment]
> 👋 friends! Over at the gutenberg-mobile repo, we seem to [run into check failures](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/12448/workflows/dae22fc0-48fe-4464-9aca-6151c5a51bf4/jobs/67087?invite=true#step-110-54) related to the JSDoc updates from this PR, and I wonder if anyone can give a quick hand or share insight. I'm sure we need to update [how we invoke the lint tests at this point](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/bin/ci-checks-js.sh#L35-L41) but any insight on what's the new proper way to do it will be appreciated! Thanks!

Hey there! Checking the Gutenberg Mobile repo, I noticed 2 things.

1) The git submodule is pointing to an older Gutenberg commit, and this old commit had a different version of the `eslint-plugin-jsdoc`.
2) The `package.json` from the root also installs the `eslint-plugin-jsdoc` in a older version.

Could it be related to these questions?

And since the `@wordpress/eslint-plugin` is coming from an older Gutenberg commit (without the jsdoc rule yet), I didn't identify quickly where it was adding this rule.

**Comment by mkevins** [comment]
Thanks for the insight @renatho and @gziolo !

> Could it be related to these questions?

Indeed it is! I updated the version of `eslint-plugin-jsdoc` in `gutenberg-mobile` to align with the version used in `gutenberg`, which seemed to resolve the issues on CI.

**Comment by hypest** [comment]
Thanks for the quick look @renatho and @gziolo ❤️ , and 👏 @mkevins for fixing the gutenberg-mobile side now 🙇

**Pull request merged**
